# <center> Logistic Regression</center>

# What is Logistic Regression

Logistic regression is a supervised classification Algorithm used when the target variable is Categorical 

Eg: yes/no
    spam/ no spam

# 1.Installing Liabary 

In [2]:
pip install pandas numpy scikit-learn matplotlib

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [5]:
pip show pandas # check the package pandas is there in the computer or not

Name: pandas
Version: 2.2.3
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: 
Author-email: The Pandas Development Team <pandas-dev@python.org>
License: BSD 3-Clause License

 Copyright (c) 2008-2011, AQR Capital Management, LLC, Lambda Foundry, Inc. and PyData Development Team
 All rights reserved.

 Copyright (c) 2011-2023, Open source contributors.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

 * Neither the name of the copyright holder nor the names of its
   contribut

In [6]:
import sys
import numpy as np
import pandas as pd
import matplotlib
import sklearn


In [8]:
df = pd.read_csv('D:/Chandra And Vennela/telecom_data.csv')
df.head()

Age  Gender PlanType  MonthlyUsage Churn
0   21  Female  Regular            15    No
1   45  Female  Economy            41    No
2   44  Female  Economy            40    No
3   31  Female  Regular            23   Yes
4   33  Female  Regular            12    No

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Age           150 non-null    int64 
 1   Gender        150 non-null    object
 2   PlanType      150 non-null    object
 3   MonthlyUsage  150 non-null    int64 
 4   Churn         150 non-null    object
dtypes: int64(2), object(3)
memory usage: 6.0+ KB


In [20]:
df.describe() # Accuracy: values realistic or outlier

Age  MonthlyUsage
count  150.000000    150.000000
mean    35.193333     33.693333
std     10.841566     15.923031
min     19.000000      3.000000
25%     25.000000     23.000000
50%     35.000000     35.000000
75%     44.000000     50.000000
max     54.000000     59.000000

In [18]:
df.shape

(150, 5)

In [14]:
df.dtypes

Age              int64
Gender          object
PlanType        object
MonthlyUsage     int64
Churn           object
dtype: object

In [25]:
df.isnull().sum() #Completeness

Age             0
Gender          0
PlanType        0
MonthlyUsage    0
Churn           0
dtype: int64

In [24]:

print(df['Gender'].value_counts(normalize=True))
print(df['Churn'].value_counts(normalize=True))

Gender
Female    0.793333
Male      0.206667
Name: proportion, dtype: float64
Churn
No     0.893333
Yes    0.106667
Name: proportion, dtype: float64


| Field | Details |
|------|--------|
| **Dataset Name** | telecom_churn_data.csv |
| **Description** | Contains customer demographic and usage details from a telecom company used to predict churn behavior. The dataset is used to demonstrate data quality checks, feature encoding, and model evaluation using Logistic Regression. |
| **Features** | Age, Gender, PlanType, MonthlyUsage |
| **Target Variable** | Churn (Yes = churned, No = retained) |
| **Number of Rows** | 150 |
| **Null Handling** | No missing values detected across any feature (`isnull().sum()` confirms all zeroes) |
| **Feature Types** | Age: Numeric (int)<br>Gender: Categorical (object, Nominal)<br>PlanType: Categorical (object, Nominal)<br>MonthlyUsage: Numeric (int)<br>Churn: Categorical (binary target) |





# 2. separate features and label

In [38]:
fdata=df.drop(columns=['Churn'])
X=fdata


y=df["Churn"].map({'Yes':1,'No':0})
categorical =fdata.select_dtypes(include='object').columns
numerical =fdata.select_dtypes(exclude='object').columns

print("Cat features:",list(categorical))
print("numerical features:",list(numerical))

Cat features: ['Gender', 'PlanType']
numerical features: ['Age', 'MonthlyUsage']


In [50]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler
X_encoded_gd = pd.get_dummies (
    X,
columns=categorical,
drop_first=True
)#Encoding directly with pandas
print("gd Encoded Columns: ",X_encoded_gd.columns.tolist())
print("------------------------------------------------------")
#print(X_encoded_gd['PlanType_Economy'].value_count())
print(X_encoded_gd.Gender_Male)

gd Encoded Columns:  ['Age', 'MonthlyUsage', 'Gender_Male', 'PlanType_Economy', 'PlanType_Regular', 'PlanType_Standard', 'PlanType_Ultra']
------------------------------------------------------
0      False
1      False
2      False
3      False
4      False
       ...  
145    False
146    False
147    False
148    False
149    False
Name: Gender_Male, Length: 150, dtype: bool


# OneHotEncoder()

In [54]:
ohe= OneHotEncoder(
drop='first',
sparse_output=False
)
encoded_array= ohe.fit_transform(X[categorical])

#Convert back to DataFrame

encoded_df = pd.DataFrame (
encoded_array,
columns=ohe.get_feature_names_out(categorical)
)

print(encoded_array)

[[0. 0. 1. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]
 [1. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 1. 0.]
 [1. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [1. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1.]
 [1. 0. 1. 0. 0.]
 [1. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 1. 0.]
 [1. 1. 0. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 1. 0. 0.]
 [1. 0. 0. 0. 1.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0.]
 [1. 0. 0. 1. 0.]
 [0. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [1. 0. 0.

In [59]:
x_encoder_ohe=pd.concat(
    [X[numerical].reset_index(drop=True),encoded_df.reset_index(drop=True)],
    axis=1
)
print("ohe:",x_encoder_ohe.columns.tolist())
print(x_encoder_ohe.head())

ohe: ['Age', 'MonthlyUsage', 'Gender_Male', 'PlanType_Economy', 'PlanType_Regular', 'PlanType_Standard', 'PlanType_Ultra']
   Age  MonthlyUsage  Gender_Male  PlanType_Economy  PlanType_Regular  \
0   21            15          0.0               0.0               1.0   
1   45            41          0.0               1.0               0.0   
2   44            40          0.0               1.0               0.0   
3   31            23          0.0               0.0               1.0   
4   33            12          0.0               0.0               1.0   

   PlanType_Standard  PlanType_Ultra  
0                0.0             0.0  
1                0.0             0.0  
2                0.0             0.0  
3                0.0             0.0  
4                0.0             0.0  


In [62]:
scaler = StandardScaler()
X_scaled_gd=X_encoded_gd.copy()
X_scaled_gd[numerical] = scaler.fit_transform(X_scaled_gd[numerical])
print(X_scaled_gd)

          Age  MonthlyUsage  Gender_Male  PlanType_Economy  PlanType_Regular  \
0   -1.313545     -1.177914        False             False              True   
1    0.907574      0.460411        False              True             False   
2    0.815027      0.397399        False              True             False   
3   -0.388079     -0.673814        False             False              True   
4   -0.202986     -1.366951        False             False              True   
..        ...           ...          ...               ...               ...   
145 -1.498638     -1.934064        False             False             False   
146  0.629934      1.153549        False             False             False   
147  1.740493     -1.682014        False             False             False   
148 -1.406091      1.468612        False              True             False   
149 -0.388079     -0.673814        False             False             False   

     PlanType_Standard  PlanType_Ultra 

In [73]:
X_scaled_ohe = X_encoded_ohe.copy()
X_scaled_ohe[numerical] = scaler.fit_transform(X_scaled_ohe[numerical]) 
df=pd.DataFrame({
'Age_ohe':X_scaled_ohe[numerical]['Age'],
'Age_gd':X_scaled_gd[numerical]['Age'],
'MonthlyUsage_ohe':X_scaled_ohe[numerical][ 'MonthlyUsage'],
'Monthleusage_gd':X_scaled_gd[numerical]['MonthlyUsage']
})
print("Comparison_Table:\n", df)

NameError: name 'X_encoded_ohe' is not defined